package.json helper class

dpogue opened the pull request:
### Platforms affected
Tooling

### What does this PR do?
Adds a class to manage finding and updating the `cordova` section of package.json, while also being consistent with npm about how newlines and indentation are handled.

The idea is to use this class in refactoring some pieces of cordova-lib.

I also added JSDoc/TypeScript-style doc comments to the class, which will provide autocompletion and tooltip information for anyone using VS Code. This partially helps to more clearly define what is "public"/private API, and also provides the opportunity to do type checking as a future testing step.

### What testing has been done on this change?
Unit tests written with 100% coverage.

brodybits commented on src/PackageHelper.js:
```diff
@@ -0,0 +1,193 @@
+const LF = '\n';
+
+// Taken from https://github.com/npm/cli/blob/ab3c62aa83375a7d4a21e396129ae5679ee86646/lib/utils/stringify-package.js
+function stringifyPackage (data, indent, newline) {
```
Why not extract stringify-package into its own package (with NPM license)?

dpogue replied:
I don't know that I feel entirely comfortable taking ownership of this as its own module, given that I didn't write any of it, but I'll ping some of the npm folk and see if they'd be interested in doing that. I'm happy to do the work of pulling it into its own module, but I'd prefer they own it.

brodybits replied:
Very thoughtful of you @dpogue. I think the ultimate solution is that they would extract stringify-package.js into its own package and reuse it within npm itself. Otherwise I guess you may as well do it yourself and offer shared ownership to the npm folk:)

brodybits commented:
> I also added JSDoc/TypeScript-style doc comments to the class

+1

dpogue commented:
If we add support for changing packaging name and version, then yes, I'd agree. Currently though, it's intended that `addPlatform`/`addPlugin` are only called when something is actually being added for the first time to the project.

raphinesse commented:
I think I would make writes explicit either way. Reasons:

- to me that would be what I would expect
- User can decide whether they want async or sync writing
- Better support for batch updates (3 new plugins, 3 writes = nasty)

I've had a few more thoughts about this:

My first thought was that this actually solves more problems than it should. To me, modifying a package.json file while keeping its formatting intact and updating cordova specific contents of package.json seem to be different enough to warrant separate tools (Interface segregation principle). I think the very generic name is partly a symptom of that too.

The use cases for `cordova-create` could easily be satisfied by a generic cordova-unaware `PackageFileEditor`, for example. Since `name` and `version` keys live on the top level, I need nothing to navigate the package object and indeed would prefer to keep it simple and just have access to it as a plain object. A hypothetical `CordovaPackageFileEditor` could delegate to the `PackageFileEditor` for file reading and writing and focus on accessing/updating the contents of the `cordova` key.

Trying to formulate suggestions for other aspects of this class, I found that it was very hard to do so because I couldn't exactly pinpoint the scope of this class.

Some use cases that I saw covered by this class:
- [x] Modify package.json while keeping its formatting intact
- [x] Safely get the cordova object
- [x] Safely get the list of plugins/platforms
- [x] Get version information for plugins and platforms

Use cases I came up with but that weren't handled (they might be irrelevant; I just don't know):
- [ ] Safely get plugin variables
- [ ] Make any other modifications to the cordova object
- [ ] Easily get the spec of a specific platform (i.e regardless of a `cordova-` prefix)

Maybe you have a better overview of the relevance of these use cases in the current code base.

Another thought was, that we might even want to aggregate information for plugins/platforms like it is in config.xml (i.e. an object with `name`, `spec` and `variable` keys describes a plugin). Even more so, we might actually want to have an interface that consistently manages platforms and plugins in package.json AND config.xml so the programmers do not have to keep them in sync themselves.

I hope these remarks weren't too much of a stream of consciousness. I just wanted to get out some of my thoughts about this in the hopes that some of them might actually be helpful. I find it very hard to do the right thing here, which is probably at least in part due to the fact that this change intersects with the topic of package management in Cordova and the messed up state of the latter.

dpogue commented:
> Maybe you have a better overview of the relevance of these use cases in the current code base.

One problem is that this class was designed as the first step in a larger proposal that I never finished writing/formatting for review. I'll try my best to get that posted on cordova-discuss today. 😅

> I think this getter and the related platform and plugin getters are somewhat problematic. 

Maybe those getters should return readonly/frozen copies of the data, to prevent trying to assign into them. They were not intended to be used for writing.

raphinesse commented:
> Maybe those getters should return readonly/frozen copies of the data, to prevent trying to assign into them.

That's exactly what I was thinking. We would need a deepFreeze lib though.

dpogue commented:
I've updated this to return frozen data anywhere that it's not returning a simple type like a string. I've also added `getPluginVariables(pluginName)` to safely get plugin variables, as well as getters for packageID (the `name` field) and version.

I've removed the calls to write, so consumers of this API will need to remember to call `write()` or `writeSync()` after making any changes.

> we might even want to aggregate information for plugins/platforms like it is in config.xml (i.e. an object with `name`, `spec` and `variable` keys describes a plugin). Even more so, we might actually want to have an interface that consistently manages platforms and plugins in package.json AND config.xml so the programmers do not have to keep them in sync themselves.

The plan is to deprecate storing any of that information in config.xml and keep it all in package.json, which this class is intended to help with. Right now it's a giant mess of over-complicated code that tries to keep them in sync, and seems to pick which file overrules the other at random, which was done as a sort of transition period. We want to only look at package.json for dependencies going forward and keep config.xml reserved for application config stuff like preferences, icons, and splashscreens.

raphinesse commented on src/PackageHelper.js:
```diff
@@ -0,0 +1,207 @@
+            // Already have the plugin installed, but might need to merge vars
+            let existing_vars = this.pkgfile.cordova.plugins[plugin];
+
+            this.pkgfile.cordova.plugins[plugin] = Object.assign({}, existing_vars, vars);
```
I would have expected this to replace the variables, not merge them. With the current implementation we cannot remove variables. Merging would still be doable, if a little verbose:
```js
const oldVars = pkgHelper.getPluginVariables('cordova-plugin-foo')
pkgHelper.addPlugin('cordova-plugin-foo', Object.assign({}, oldVars, newVars))
```
I couldn't tell whether we have to merge often. If so, we should probably provide an option for that.

dpogue commented:
TODO: Address https://github.com/apache/cordova-lib/issues/694 here to avoid writing if there are no changes

dpogue commented:
Also possibly worth looking at: https://github.com/npm/read-package-json

dpogue commented:
I'm hoping to update this before it's merged. This probably won't make the next major

brodybits commented:
I just marked this as [WIP] to avoid a premature merge.

raphinesse commented:
`modified` should be reset after writing out to disk.

I'd personally prefer to keep a copy of the initial contents and use `deepEqual` to check if need to write to disk. That would also catch use-cases where we need to manipulate `pgkfile` directly.

---
`cordovaDependencies` should be updated to avoid problems when `dependencies` or `devDependencies` are `undefined`:

```js
get cordovaDependencies () {
    const mergedDependencies = Object.assign({}, this.pkgfile.dependencies, this.pkgfile.devDependencies);
    return deepFreeze(
        this.cordovaPlatforms
            .map(p => p.startsWith('cordova-') ? p : `cordova-${p}`)
            .concat(this.cordovaPlugins)
            .reduce((deps, cur) => {
                deps[cur] = mergedDependencies[cur];
                return deps;
            }, {})
    );
}
```

---


> we might even want to aggregate information for plugins/platforms like it is in config.xml (i.e. an object with `name`, `spec` and `variable` keys describes a plugin).

When test-driving this class on the new `restoreUtil.installPlatformsFromConfigXML` aggregate plugin information as requested above would still have been useful, even when we are not using config.xml to store dependency information anymore.

---

As this class has reached quite some size and still could do more I stand by my remarks from above and would appreciate feedback on it.

> My first thought was that this actually solves more problems than it should. To me, modifying a package.json file while keeping its formatting intact and updating cordova specific contents of package.json seem to be different enough to warrant separate tools (Interface segregation principle). I think the very generic name is partly a symptom of that too.

> The use cases for `cordova-create` could easily be satisfied by a generic cordova-unaware `PackageFileEditor`, for example. Since `name` and `version` keys live on the top level, I need nothing to navigate the package object and indeed would prefer to keep it simple and just have access to it as a plain object. A hypothetical `CordovaPackageFileEditor` could delegate to the `PackageFileEditor` for file reading and writing and focus on accessing/updating the contents of the `cordova` key.

dpogue commented:
@raphinesse Thanks for the re-review! I agree with almost all of those comments. Funny thing, when working on https://github.com/apache/cordova-lib/pull/752 I was referencing back to this PR and actually wrote it to use this class's API before refactoring it to the expanded form.

I definitely see the value in splitting this up into a PackageFileEditor and a CordovaPackageFileEditor subclass. 🙂 

The only suggestion I'm not really a fan of is changing the format of the data stored in package.json, unless you were thinking just of providing a getter that returned an object with aggregated data?

> we might even want to aggregate information for plugins/platforms like it is in config.xml (i.e. an object with `name`, `spec` and variable keys describes a plugin).

raphinesse commented:
> The only suggestion I'm not really a fan of is changing the format of the data stored in package.json, unless you were thinking just of providing a getter that returned an object with aggregated data?

Yes, my suggestion was to provide getters similar to `ConfigParser.getPlugin` and maybe `ConfigParser.getPlugins` while keeping the data format stored in package.json untouched.

https://github.com/apache/cordova-common/blob/f9951b360f8db344f919b1ee0c87c26661698a98/src/ConfigParser/ConfigParser.js#L400-L427

https://github.com/apache/cordova-common/blob/f9951b360f8db344f919b1ee0c87c26661698a98/src/ConfigParser/ConfigParser.js#L355-L359

dpogue commented:
Okay cool, that definitely sounds reasonable to do :)

piotr-cz commented:
First of all sorry for my ignorace, as I haven't read this PR's commits and comments.

I just want to say that IMHO package.json should be saved in same format as npm uses.
At the moment every npm operation adds newline at the end of the file and every cordova operation removes it.

brodybits commented:
What is the status of this?

dpogue commented:
> What is the status of this?

No change in status. I've had no time to look at this any further.
I think the next step was going to be splitting this into two classes: one that handles package.json, and one that is a subclass that handles the cordova-specific bits.



> BTW it's handled via [stringify-package](https://www.npmjs.com/package/stringify-package) package

Yep, I extracted that from npm CLI into its own module as part of putting this proposal together :)

breautek commented on package.json:
```diff
@@ -26,16 +26,22 @@
     "minimatch": "^3.0.0",
     "plist": "^3.0.1",
     "q": "^1.4.1",
+    "stringify-package": "^1.0.0",
```
this package is [deprecated](https://www.npmjs.com/package/stringify-package), the replacement is https://www.npmjs.com/package/@npmcli/package-json

erisu commented:
I think we can actually close this PR for `[@npmcli/package-json](https://www.npmjs.com/package/@npmcli/package-json)`.
